sklearn/semi_supervised/_self_training.py

```diff
@@ -176,11 +181,17 @@ def fit(self, X, y):
         self : object
             Fitted estimator.
         """
-        # we need row slicing support for sparce matrices, but costly finiteness check
-        # can be delegated to the base estimator.
-        X, y = self._validate_data(
-            X, y, accept_sparse=["csr", "csc", "lil", "dok"], force_all_finite=False
-        )
+        y = self._validate_data(X="no_validation", y=y)
+        check_consistent_length(X, y)
+        if issparse(X):
+            X = _ensure_sparse_format(
+                X,
+                accept_sparse=["csr", "csc", "lil", "dok"],
```

Comment:
This is required for `check_estimator_sparse_data` to pass. `_safe_indexing` does not work with the `dia_matrix` sparse format.